### PR TITLE
Try py3 compatibility

### PR DESCRIPTION
**Pull request by isaacm**
Please review.

### CODE REVIEW

File: moecache.py [hunk 1]
```diff
@@ -122,7 +134,7 @@ def connect(self):
             # thus, might read more than the current expected response
             # cleared on every reconnect since old bytes are part of old
             # session and can't be reused
-            self._buffer = ''
+            self._buffer = bytearray('', 'utf-8')
```
Comment (lichray) [inline moecache.py]: Why not all use `b''`?

File: moecache.py [hunk 2]
```diff
@@ -460,20 +485,22 @@ def stats(self, additional_args=None):
         # resp - STAT <name> <value>\r\n (one per result)
         #        END\r\n
         if additional_args is not None:
-            command = 'stats ' + additional_args + '\r\n'
+            command = bytearray('stats ', 'utf-8') + \
+                bytearray(additional_args, 'utf-8') + \
```
Comment (lichray) [inline moecache.py]: Leave user's argument as-is.

File: moecache.py [hunk 3]
```diff
@@ -353,25 +371,32 @@ def get(self, key):
...
                 typecode = int(terms[2]) ^ 0x100
                 length = int(terms[3])
                 if typecode > 0xff:
                     error = ClientException('not a moecache deployment')
-                if terms[1] == key:
+                if terms[1] == bytearray(key, 'utf-8'):
                     received = node.gets(length+2)[:-2]
                     if typecode == 18:
-                        val = received
+                        val = received.decode('utf-8')
```
Comment (lichray) [inline moecache.py]: `val`'s type is changed to `unicode` in Python2, thus a potential break.  I suggest to leave it as is.

File: moecache.py [hunk 4]
```diff
@@ -460,20 +485,22 @@ def stats(self, additional_args=None):
...
         def do_stats(node):
             resp = node.send(command)
             result = {}
-            while resp != 'END\r\n':
+            while resp != bytearray('END\r\n', 'utf-8'):
                 terms = resp.split()
-                if len(terms) == 3 and terms[0] == 'STAT':
-                    result[terms[1]] = terms[2]
+                if len(terms) == 3 and terms[0] == bytearray('STAT', 'utf-8'):
+                    result[terms[1].decode('utf-8')] = terms[2].decode('utf-8')
```
Comment (lichray) [inline moecache.py]: Leave it as-is.

File: moecache.py (hunk 1 above)
Comment (lichray) [inline moecache.py]: Oh, buffer suppose to be modifiable.  But other literals should be just bytes.  Also drop the `'utf-8'` argument, not needed for non-Unicode literals.

File: tests/mock_memcached.py [hunk 5]
```diff
@@ -110,15 +112,15 @@ def run(self):
...
                 terms = request.split()
-                if len(terms) == 2 and terms[0] == 'get':
+                if len(terms) == 2 and terms[0] == b'get':
                     self._handle_get(terms[1])
-                elif len(terms) == 5 and terms[0] == 'set':
+                elif len(terms) == 5 and terms[0] == b'set':
                     self._handle_set(terms[1], int(terms[4]))
                 else:
-                    print 'unknown command', repr(request)
+                    print('unknown command', repr(request.decode('utf-8')))
```
Comment (lichray) [inline tests/mock_memcached.py]: no need to decode, `repr` gives you string.

File: moecache.py [hunk 6]
```diff
@@ -333,8 +347,12 @@ def delete(self, key):
         self._validate_key(key)
 
         command = 'delete ' + key + '\r\n'
+        command = bytearray(command, 'utf-8')
```
Comment (lichray) [inline moecache.py]: Use `bytes` as key, and create `command` as bytes in the first place.

**Comment by isaacm** [comment]
I'll address shortly.

Side note, when working on this library were you able to pass all [tests](https://github.com/lichray/moecache/blob/master/tests/test.py) locally?

**Comment by lichray** [comment]
Yes.  The coverage is ~87% iirc, but all tests pass.

File: moecache.py [hunk 7]
```diff
@@ -353,25 +371,32 @@ def get(self, key):
...
-                if terms[1] == key:
+                if terms[1] == bytearray(key, 'utf-8'):
                     received = node.gets(length+2)[:-2]
                     if typecode == 18:
-                        val = received
+                        val = received.decode('utf-8')
                     elif typecode == 0:
-                        val = pickle_loads(received)
+                        if is_py2:
+                            val = pickle_loads(str(received))
```
Comment (lichray) [inline moecache.py]: Why we need this?  `pickle.loads` works with bytes in both versions.

**Comment by lichray** [comment]
Suggestion: make the library accept both bytes and unicode string in python3, and returns unicode string in python3.  `stats()` is kept bytes-only.

File: moecache.py (hunk 1 above)
Comment (lichray) [inline moecache.py]: `self._buffer = bytearray()`

File: moecache.py [hunk 8]
```diff
@@ -55,13 +56,24 @@ def pickle_dumps(obj):
...
+        from io import BytesIO as OStringStream
+        import pickle
+
+        def pickle_dumps(obj):
+            return pickle.dumps(obj, 2)
+
+        pickle_loads = pickle.loads
 
-    pickle_loads = cPickle.loads
+is_py2 = (sys.version_info[0] == 2)
```
Comment (lichray) [inline moecache.py]: Make this a static private class member instead of global, and detect python 3 not 2.

File: moecache.py [hunk 9]
```diff
@@ -122,7 +132,10 @@ def connect(self):
             # thus, might read more than the current expected response
             # cleared on every reconnect since old bytes are part of old
             # session and can't be reused
-            self._buffer = ''
+            if Client.is_py3:
```
Comment (lichray) [inline moecache.py]: I guess we can use `bytearray` for python2 as well.

File: moecache.py [hunk 10]
```diff
@@ -330,11 +359,21 @@ def delete(self, key):
...
+        key_bytes = self._validate_key(key)
+
+        if Client.is_py3:
+            command = (
+                bytearray('delete ', 'utf-8') +
+                key_bytes +
+                bytearray('\r\n', 'utf-8')
+            )
+        else:
+            command = 'delete ' + key + '\r\n'
```
Comment (lichray) [inline moecache.py]: Just say `b'delete ' + key + b'\r\n'`, works for both versions.

File: moecache.py [hunk 11]
```diff
@@ -350,28 +389,45 @@ def get(self, key):
...
+                        if Client.is_py3:
+                            val = received.decode('utf-8')
+                        else:
+                            val = received
                     elif typecode == 0:
-                        val = pickle_loads(received)
+                        if Client.is_py3:
+                            val = pickle_loads(received)
+                        else:
+                            val = pickle_loads(str(received))
```
Comment (lichray) [inline moecache.py]: Why we need this?  `pickle` of course consumes bytes in python2.

File: moecache.py [hunk 12]
```diff
@@ -422,25 +478,39 @@ def set(self, key, val, exptime=0):
 
         if isinstance(val, str):
             flag = ' 274 '  # 18 | 0x100
-            sent = val
+            if Client.is_py3:
+                sent = bytearray(val, 'utf-8')
```
Comment (lichray) [inline moecache.py]: `val.decode('utf-8')` is more traditional.

File: moecache.py [hunk 13]
```diff
@@ -422,25 +478,39 @@ def set(self, key, val, exptime=0):
...
+            buf.write(bytearray(' ', 'utf-8'))
+            buf.write(bytearray(str(len(sent)), 'utf-8'))
+            buf.write(bytearray('\r\n', 'utf-8'))
+            buf.write(sent)
+            buf.write(bytearray('\r\n', 'utf-8'))
+        else:
+            buf.write('set ')
+            buf.write(key)
+            buf.write(flag)
+            buf.write(str(exptime))
```
Comment (lichray) [inline moecache.py]: The whole switch is not needed.  `b` prefix works in python2.  For `str(exptime)`, `str(exptime).encode()` works in both versions.

File: moecache.py (hunk 12 above)
Comment (isaacm) [inline moecache.py]: I think you meant `val.encode('utf-8')` , changed it to encode.

**Pull request closed without merging**